File: javascript/allergies/allergies.js
var allergenScores = {
  "cats": 128,
  "pollen": 64,
  "chocolate": 32,
  "tomatoes": 16,
  "strawberries": 8,
  "shellfish": 4,
  "peanuts": 2,
  "eggs": 1
};

function Allergies(score) {
  this.list = function() {
    var list = [];
    Object.keys(allergenScores).forEach(function(key) {
      score %= 256;
      if (score > 0 && score / allergenScores[key] >= 1) {
        list.unshift(key);
        score -= allergenScores[key];
      }
    });
    return list;
  };
  this.allergicTo = function(allergen) {
    return score > 0 && score % allergenScores[allergen] === 0;
  };
};

module.exports = Allergies;

File: csharp/allergies/Allergies.cs
using System.Collections.Generic;

public class Allergies
{ 
  private Dictionary<string, int> AllergenScores = new Dictionary<string, int>()
  {
    { "eggs", 1 },
    { "peanuts", 2 },
    { "shellfish", 4 },
    { "strawberries", 8 },
    { "tomatoes", 16 },
    { "chocolate", 32 },
    { "pollen", 64 },
    { "cats", 128 }
  };

  private int score;

  public Allergies(int score)
  {
    this.score = score;
  }

  public bool AllergicTo(string allergen)
  {
    return isAllergic(AllergenScores[allergen]);
  }

  public List<string> List()
  {
    List<string> allergies = new List<string>();
    foreach (var pair in AllergenScores)
    {
      if (isAllergic(pair.Value)) 
      {
        allergies.Add(pair.Key);
      }
    }
    return allergies;
  }

  private bool isAllergic(int allergen)
  {
    return (score & allergen) == allergen;
  }
}
